powerscale/provider/snapshot_resource.go

```diff
@@ -0,0 +1,326 @@
+		)
+
+		return
+	}
+
+	r.client = pscaleClient
+}
+
+// Create allocates the resource.
+func (r *SnapshotResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	tflog.Info(ctx, "creating snapshot")
+	var plan *models.SnapshotDetailModel
+	// Read Terraform plan data into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+
+	if resp.Diagnostics.HasError() {
+
+		return
+	}
+	result, err := helper.CreateSnapshot(ctx, r.client, plan)
+	if err != nil {
+		errStr := constants.CreateSnapshotErrorMessage + " with error: "
+		message := helper.GetErrorString(err, errStr)
+		resp.Diagnostics.AddError(
+			"Error creating snapshot",
+			message,
+		)
+		return
+	}
+
+	state, err := helper.SnapshotResourceDetailMapper(ctx, result)
+	if err != nil {
+		resp.Diagnostics.AddError(constants.CreateSnapshotErrorMessage+" with error: ", err.Error())
+		return
+	}
+	state.SetExpires = plan.SetExpires
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+}
+
+// Read reads the resource state.
+func (r *SnapshotResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	tflog.Info(ctx, "reading snapshot")
+	var plan *models.SnapshotDetailModel
+
+	// Read Terraform prior state plan into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &plan)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	result, err := helper.GetSpecificSnapshot(ctx, r.client, plan.ID.ValueString())
+	if err != nil {
+		errStr := constants.ReadSnapshotErrorMessage + " with error: "
+		message := helper.GetErrorString(err, errStr)
+		resp.Diagnostics.AddError(
+			"Error getting the list of snapshots",
+			message,
+		)
+		return
+	}
+
+	state, err := helper.SnapshotResourceDetailMapper(ctx, result)
+	if err != nil {
+		resp.Diagnostics.AddError(constants.ReadSnapshotErrorMessage+" with error: ", err.Error())
+		return
+	}
+	state.SetExpires = plan.SetExpires
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+}
+
+// Update updates the resource state Path, Name, AuthProviders.
+func (r *SnapshotResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	tflog.Info(ctx, "updating snapshot")
+	var plan *models.SnapshotDetailModel
+	var state *models.SnapshotDetailModel
+
+	// Read Terraform plan data into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Read the state
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if plan.Path != state.Path {
+		resp.Diagnostics.AddError(
+			"Error editing snapshot",
+			"Path variable is not able to be updated after creation",
+		)
+		return
+	}
+
+	// Populate the Edit Parameters
+	editVal := powerscale.V1SnapshotSnapshotExtendedExtended{}
+	if plan.Name != state.Name {
+		editVal.Name = plan.Name.ValueStringPointer()
+		state.Name = plan.Name
```

Comment:
@doriac11 Why are we updating the state with plan data even before modify request is sent? Ideally state should have been updated only once modify is successful.

Reply:
Me and @nikitajoshi1 discussed in slack, this solution should be suitable.